codecs/jxl/dec/jxl_dec.cpp

```diff
@@ -1,31 +1,77 @@
 #include <emscripten/bind.h>
 #include <emscripten/val.h>
 
-#include <jpegxl/decode.h>
+#include <jxl/decode.h>
+#include "lib/jxl/color_encoding_internal.h"
+
+#include "skcms.h"
 
 using namespace emscripten;
 
 thread_local const val Uint8ClampedArray = val::global("Uint8ClampedArray");
 thread_local const val ImageData = val::global("ImageData");
 
-#define EXPECT_EQ(a, b) if ((a) != (b)) { JpegxlDecoderDestroy(dec); return val::null(); }
+#define EXPECT_EQ(a, b)     \
+  if ((a) != (b)) {         \
+    JxlDecoderDestroy(dec); \
+    return val::null();     \
+  }
 
 val decode(std::string data) {
-  JpegxlDecoder* dec = JpegxlDecoderCreate(nullptr);
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderSubscribeEvents(dec, JPEGXL_DEC_BASIC_INFO | JPEGXL_DEC_FULL_IMAGE));
+  JxlDecoder* dec = JxlDecoderCreate(nullptr);
+  EXPECT_EQ(JXL_DEC_SUCCESS,
+            JxlDecoderSubscribeEvents(
+                dec, JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING | JXL_DEC_FULL_IMAGE));
   auto next_in = (const uint8_t*)data.c_str();
   auto avail_in = data.size();
-  EXPECT_EQ(JPEGXL_DEC_BASIC_INFO, JpegxlDecoderProcessInput(dec, &next_in, &avail_in));
+  EXPECT_EQ(JXL_DEC_BASIC_INFO, JxlDecoderProcessInput(dec, &next_in, &avail_in));
   size_t buffer_size;
-  const JpegxlPixelFormat format = {4, JPEGXL_LITTLE_ENDIAN, JPEGXL_TYPE_UINT8};
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderImageOutBufferSize(dec, &format, &buffer_size));
-  JpegxlBasicInfo info;
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderGetBasicInfo(dec, &info));
+  const JxlPixelFormat format = {4, JXL_LITTLE_ENDIAN, JXL_TYPE_UINT8};
```

Comment:
Using output type uint8 will cause clamping if the image was not originally sRGB, is this intended? If not, float is better. uint16 will also be clamped

Reply:
Squoosh uses uint8 anyway (possibly because that's what Canvas expects?), so I don't think I can do this properly with floats.
Floats were also causing me an out-of-memory error (or a segfault), which is at least a bit problematic.

Reply:
you're using skcms first though, right?

the correct method is:

decode to float --> convert with skcms to the correct color space --> convert to uint8 or uint16 and clamp

the clamping should not be done before the conversion with skcms (it throws away colors).

will that work?

Reply:
The web platform — as of now — only supports `uint8` for [`ImageData`](https://developer.mozilla.org/en-US/docs/Web/API/ImageData). We have to figure something out here eventually if we want to support HDR, but for now, yes, we are clamping (or scaling) to fit everything into `uint8`.

Reply:
I'm using floats as intermediate values now - my previous attempt used the buffer size instead of the number of pixels, explaining the segmentation fault...